test/functional/eval/capture_spec.lua

```diff
@@ -0,0 +1,39 @@
+local helpers = require('test.functional.helpers')
+local eq = helpers.eq
+local eval = helpers.eval
+local clear = helpers.clear
+local execute = helpers.execute
+local source = helpers.source
+
+describe('capture()', function()
+  before_each(clear)
+
+  it('returns the same result with :redir', function()
+    execute('redir => g:foo | messages | redir END')
+    eq(eval('g:foo'), eval('capture(":messages")'))
+  end)
+
+  it('returns the output of the commands if the argument is List', function()
+    eq("foobar", eval([[capture(['echon "foo"', 'echon "bar"'])]]))
+  end)
+
+  it('supports the nested redirection', function()
+    source([[
+    function! g:Foo()
+      let a = ''
+      redir => a
+      silent echon "foo"
+      redir END
+      return a
+    endfunction
+    function! g:Bar()
+      let a = ''
+      redir => a
+      call g:Foo()
+      redir END
+      return a
+    endfunction
+    ]])
+    eq('foo', eval('capture("call g:Bar()")'))
+  end)
```

Comment:
Missing tests:

1. Nested `capture('echo capture("echo 42")')`.
2. Nested `redir => g:mes|echo capture('echo capture("redir => g:mes2|echo 42|redir END")')|redir END` (check `g:mes` and `g:mes2`). No functions!
3. `capture([])`
4. `capture(v:_null_list)`
5. Errors: `capture(0.0)`, `capture(v:_null_dict)`, `capture(function('tr'))` (use `exc_exec`).
6. Errors: same, but inside a list. Better check separately when invalid type has neither first nor last element of the list. Here  and previously `redir_exec` with `echo capture()` may be used if you think that output of `capture()` in case of error should be something specific, otherwise `exc_exec`.